libqtile/popup/menu.py

```diff
@@ -0,0 +1,174 @@
+            offset = self.icon_size + self.icon_gap
+
+        self.drawer.ctx.save()
+        self.drawer.ctx.translate(offset,
+                                  int((self.height - self.layout.height) / 2))
+        self.layout.draw(0, 0)
+        self.drawer.ctx.restore()
+
+
+class PopupMenuSeparator(PopupSlider):
+    defaults = [
+        ("colour_above", "ff00ff", "Separator colour"),
+        ("end_margin", 10, ""),
+        ("marker_size", 0, "")
+    ]
+
+    def __init__(self, **config):
+        PopupSlider.__init__(self, value=0, **config)
+        self.add_defaults(PopupMenuSeparator.defaults)
+
+
+class PopupMenu(PopupGridLayout):
+
+    def __init__(self, qtile, controls, **config):
+        PopupGridLayout.__init__(self, qtile, controls=controls, **config)
+        self._hide_timer = None
+
+    def process_pointer_enter(self, x, y):
+        PopupGridLayout.process_pointer_enter(self, x, y)
+        if self._hide_timer is not None:
+            self._hide_timer.cancel()
+            self._hide_timer = None
+
+    def process_pointer_leave(self, x, y):
+        PopupGridLayout.process_pointer_leave(self, x, y)
+        self._hide_timer = self.qtile.call_later(0.5, self.kill)
+
+    @classmethod
+    def from_dbus_menu(cls, qtile, dbusmenuitems, **config):
+        menuitems = []
+        prev_sep = False
+        row = 0
+        for i, dbmitem in enumerate(dbusmenuitems):
+            label = dbmitem.label.replace("_", "")
+            sep = dbmitem.item_type == "separator"
+            if not dbmitem.visible:
+                continue
+
+            if prev_sep and sep:
+                continue
+
+            if sep:
+                menuitems.append(
+                    PopupMenuSeparator(
+                        row=row,
+                        rowspan=1,
+                        bar_size=1,
+                        colour_above=config["separator_colour"]
+                    )
+                )
+            else:
+                menuitems.append(
+                    PopupMenuItem(
+                        row=row,
+                        col=0,
+                        row_span=2,
+                        text=label,
+                        menu_icon=dbmitem.icon_data,
+                        hover=True,
+                        mouse_callbacks={"Button1": lambda dbmitem=dbmitem: dbmitem.click()},
+                        font_size=config["fontsize"],
+                        foreground=config["foreground"],
+                        highlight=config["highlight"],
+                        show_icon=config["show_menu_icons"]
+                    )
+                )
+            prev_sep = sep
+            row += 1 if sep else 2
+
+        return cls.generate(qtile, menuitems, **config)
+
+    @classmethod
+    def generate(cls, qtile, menuitems, **config):
+        rowcount = sum(x.row_span for x in menuitems)
+        menu_config = {
+            "width": 200,
+            "height": 12 * rowcount + 10,
```

Comment:
Still hardcoded. This should vary based on fontsize. Is there a formula for this?